src/adm.rs

```diff
@@ -1,39 +1,269 @@
 use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
 use url::Url;
 
-use crate::error::HandlerError;
+use crate::error::{HandlerError, HandlerErrorKind, HandlerResult};
+use crate::server::location::LocationResult;
+use crate::server::ServerState;
+use crate::settings::Settings;
+use crate::tags::Tags;
+use crate::web::middleware::sentry as l_sentry;
+//use crate::server::img_storage;
+
+pub(crate) const DEFAULT: &str = "DEFAULT";
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct AdmTileResponse {
     pub tiles: Vec<AdmTile>,
 }
 
-#[derive(Debug, Deserialize, Serialize)]
+/// Filter criteria for adm Tiles
+#[derive(Default, Clone, Debug)]
+pub struct AdmFilter {
+    /// list of allowed base host strings.
+    pub filter_set: BTreeMap<String, AdmAdvertiserFilterSettings>,
+}
+
+/// The AdmAdvertiserFilterSettings contain the settings for the various
+/// ADM provided partners. These are specified as a JSON formatted hash
+/// that contains the components. A special "DEFAULT" setting provides
+/// information that may be used as a DEFAULT, or commonly appearing set
+/// of data.
+/// See `impl From<Settings>` for details of the structure.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct AdmAdvertiserFilterSettings {
+    pub(crate) advertiser_hosts: Vec<String>,
+    pub(crate) impression_hosts: Vec<String>,
+    pub(crate) click_hosts: Vec<String>,
+    pub(crate) position: Option<u8>,
+    pub(crate) include_regions: Vec<String>,
+}
+
+pub(crate) type AdmSettings = HashMap<String, AdmAdvertiserFilterSettings>;
+
+macro_rules! check_url {
+    ($url:expr, $species:expr, $filter:expr, $tags: expr) => {{
+        let parsed: Url = match $url.parse() {
+            Ok(v) => v,
+            Err(e) => {
+                $tags.add_tag("type", $species);
+                $tags.add_extra("parse_error", &e.to_string());
+                $tags.add_extra("url", $url);
+                return Err(HandlerErrorKind::InvalidHost($species, $url.to_string()).into());
+            }
+        };
+        let host = match parsed.host() {
+            Some(v) => v.to_string(),
+            None => {
+                $tags.add_tag("type", $species);
+                $tags.add_extra("url", $url);
+                return Err(HandlerErrorKind::MissingHost($species, parsed.to_string()).into());
+            }
+        };
+        if !$filter.contains(&host) {
+            $tags.add_tag("type", $species);
+            $tags.add_extra("url", $url);
+            return Err(HandlerErrorKind::UnexpectedHost($species, host).into());
+        }
+        Ok(())
+    }};
+}
```

Comment:
Apologies if I'm missing something obvious, but why is this a macro instead of a function like this?

```rust
fn check_url(url: &str, species: &str, filter: &AdmAdvertiserFilterSettings, tags: &Tags) -> HandlerResult<()>
```

Reply:
because when I started i was doing more things that couldn't be done with just a function call (e.g. inlining the `$species` into the returned error names), and was change blind to the fact that it didn't need to be a macro anymore.

I'll convert it to a full function, thanks!